File: analyzer.py
"""Analyze upgrade PRs to determine status and needed actions."""

import logging
import re
from dataclasses import dataclass, field
from enum import Enum

import gh
from config import AgentConfig

log = logging.getLogger(__name__)


class PRStatus(Enum):
    WAITING_FOR_CI = "waiting_for_ci"
    CI_FAILING = "ci_failing"
    CI_PASSED = "ci_passed"
    HAS_REVIEW_ISSUES = "has_review_issues"
    READY_TO_APPROVE = "ready_to_approve"
    APPROVED = "approved"
    READY_TO_MERGE = "ready_to_merge"
    MERGED = "merged"
    CLOSED = "closed"
    SUPERSEDED = "superseded"
    UNKNOWN = "unknown"


@dataclass
class FixableIssue:
    description: str
    file_path: str
    fix_type: str  # "version_mismatch", "lockfile_stale", "regen_needed"
    details: dict = field(default_factory=dict)


@dataclass
class PRAnalysis:
    number: int
    title: str
    head_sha: str
    head_ref: str
    status: PRStatus
    ci_passed: bool = False
    ci_pending: bool = False
    ci_failed_checks: list[str] = field(default_factory=list)
    has_approvals: bool = False
    approval_count: int = 0
    fixable_issues: list[FixableIssue] = field(default_factory=list)
    unfixable_issues: list[str] = field(default_factory=list)
    is_superseded: bool = False
    newer_pr_number: int | None = None
    mergeable: bool = False


def analyze_pr(config: AgentConfig, pr_number: int,
               all_open_prs: list[dict] | None = None) -> PRAnalysis:
    """Fully analyze an upgrade PR and determine what action to take."""
    pr = gh.get_pr(config.owner, config.repo, pr_number)

    if pr.get("merged"):
        return PRAnalysis(
            number=pr_number, title=pr["title"],
            head_sha=pr["head"]["sha"], head_ref=pr["head"]["ref"],
            status=PRStatus.MERGED,
        )
    if pr.get("state") == "closed":
        return PRAnalysis(
            number=pr_number, title=pr["title"],
            head_sha=pr["head"]["sha"], head_ref=pr["head"]["ref"],
            status=PRStatus.CLOSED,
        )

    analysis = PRAnalysis(
        number=pr_number,
        title=pr["title"],
        head_sha=pr["head"]["sha"],
        head_ref=pr["head"]["ref"],
        status=PRStatus.UNKNOWN,
        mergeable=pr.get("mergeable_state") == "clean",
    )

    # Check if superseded by a newer PR with the same pattern
    _check_superseded(config, analysis, pr, all_open_prs)
    if analysis.is_superseded:
        analysis.status = PRStatus.SUPERSEDED
        return analysis

    # Check CI status
    _check_ci(config, analysis)

    # Check reviews and comments for issues
    _check_reviews(config, analysis)

    # Determine final status
    # Fixable issues take priority over CI failure — the failure may be
    # *caused* by the issues (e.g. stale lockfile → npm ci mismatch).
    if analysis.fixable_issues:
        analysis.status = PRStatus.HAS_REVIEW_ISSUES
    elif analysis.ci_pending:
        analysis.status = PRStatus.WAITING_FOR_CI
    elif not analysis.ci_passed:
        analysis.status = PRStatus.CI_FAILING
    elif analysis.has_approvals:
        analysis.status = PRStatus.READY_TO_MERGE
    else:
        analysis.status = PRStatus.READY_TO_APPROVE

    return analysis


def _check_superseded(config: AgentConfig, analysis: PRAnalysis,
                      pr: dict, all_open_prs: list[dict] | None):
    """Check if a newer PR with the same upgrade pattern exists."""
    if not all_open_prs:
        return

    # Extract version pattern from title, e.g. "Update UnbrandedGeneratorVersion to X"
    title = pr["title"]
    for pattern in config.pr_title_patterns:
        if pattern in title:
            # Find other open PRs with the same prefix
            same_type = [
                p for p in all_open_prs
                if pattern in p.get("title", "")
                and p["number"] != analysis.number
                and p["number"] > analysis.number  # newer = higher number
            ]
            if same_type:
                newest = max(same_type, key=lambda p: p["number"])
                analysis.is_superseded = True
                analysis.newer_pr_number = newest["number"]
                log.info("PR #%d is superseded by #%d",
                         analysis.number, newest["number"])
            break


def _check_ci(config: AgentConfig, analysis: PRAnalysis):
    """Check CI check-run status on the head commit."""
    checks = gh.get_check_runs(config.owner, config.repo, analysis.head_sha)

    if not checks:
        analysis.ci_pending = True
        return

    # Find the main CI check (or evaluate all)
    main_check = None
    for check in checks:
        name = check.get("name", "")
        if name == config.required_check_pattern:
            main_check = check
            break

    if main_check:
        status = main_check.get("status")
        conclusion = main_check.get("conclusion")
        if status != "completed":
            analysis.ci_pending = True
        elif conclusion == "success":
            analysis.ci_passed = True
        else:
            analysis.ci_passed = False
            analysis.ci_failed_checks.append(
                f"{main_check['name']}: {conclusion}"
            )
    else:
        # Evaluate all checks
        completed = [c for c in checks if c.get("status") == "completed"]
        pending = [c for c in checks if c.get("status") != "completed"]
        failed = [c for c in completed if c.get("conclusion") != "success"]

        if pending:
            analysis.ci_pending = True
        elif failed:
            analysis.ci_passed = False
            for c in failed:
                analysis.ci_failed_checks.append(
                    f"{c['name']}: {c.get('conclusion')}"
                )
        else:
            analysis.ci_passed = True


def _check_reviews(config: AgentConfig, analysis: PRAnalysis):
    """Check reviews and review comments for issues and approvals."""
    reviews = gh.get_pr_reviews(config.owner, config.repo, analysis.number)
    comments = gh.get_pr_review_comments(
        config.owner, config.repo, analysis.number
    )

    # Count approvals
    approvals = [r for r in reviews if r.get("state") == "APPROVED"]
    analysis.approval_count = len(approvals)
    analysis.has_approvals = len(approvals) > 0

    # Analyze review comments for fixable issues
    for comment in comments:
        body = comment.get("body", "")
        path = comment.get("path", "")
        _classify_review_comment(analysis, body, path)


def _classify_review_comment(analysis: PRAnalysis, body: str, path: str):
    """Classify a review comment as fixable or not."""
    body_lower = body.lower()

    # Pattern: package-lock.json out of sync
    if ("package-lock" in body_lower and
            ("stale" in body_lower or "still pins" in body_lower or
             "regenerate" in body_lower or "update" in body_lower)):
        analysis.fixable_issues.append(FixableIssue(
            description="package-lock.json is out of sync with package.json",
            file_path=path,
            fix_type="lockfile_stale",
        ))
        return

    # Pattern: version mismatch between files
    if ("bump" in body_lower or "version" in body_lower) and (
            "mismatch" in body_lower or "still" in body_lower or
            "please" in body_lower):
        # Extract version info if possible
        version_match = re.search(r'(\d+\.\d+\.\d+(?:-[a-zA-Z0-9.]+)?)',
                                  body)
        analysis.fixable_issues.append(FixableIssue(
            description=f"Version needs updating in {path}",
            file_path=path,
            fix_type="version_mismatch",
            details={
                "suggested_version": version_match.group(1)
                if version_match else None,
            },
        ))
        return

    # Pattern: devDependencies not bumped
    if "devdependencies" in body_lower or "dev dependencies" in body_lower:
        analysis.fixable_issues.append(FixableIssue(
            description="devDependencies version needs bumping",
            file_path=path,
            fix_type="version_mismatch",
        ))
        return

    # If we can't classify it, it's unfixable (needs human review)
    if len(body) > 50:  # Ignore trivial bot messages
        analysis.unfixable_issues.append(
            f"[{path}] {body[:200]}"
        )

File: Analyzer.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace UpgradePrAgent;

public enum PrStatus
{
    WaitingForCi,
    CiFailing,
    CiPassed,
    HasReviewIssues,
    ReadyToApprove,
    Approved,
    ReadyToMerge,
    Merged,
    Closed,
    Superseded,
    Unknown,
}

public sealed record FixableIssue(string Description, string FilePath, string FixType,
    Dictionary<string, string?>? Details = null);

public sealed class PrAnalysis
{
    public int Number { get; init; }
    public string Title { get; init; } = "";
    public string HeadSha { get; init; } = "";
    public string HeadRef { get; init; } = "";
    public PrStatus Status { get; set; } = PrStatus.Unknown;
    public bool CiPassed { get; set; }
    public bool CiPending { get; set; }
    public List<string> CiFailedChecks { get; } = [];
    public bool HasApprovals { get; set; }
    public int ApprovalCount { get; set; }
    public List<FixableIssue> FixableIssues { get; } = [];
    public List<string> UnfixableIssues { get; } = [];
    public bool IsSuperseded { get; set; }
    public int? NewerPrNumber { get; set; }
    public bool Mergeable { get; set; }
    public bool HasMergeConflicts { get; set; }
}

public static partial class Analyzer
{
    public static async Task<PrAnalysis> AnalyzePrAsync(
        AgentConfig config, int prNumber, List<JsonElement>? allOpenPrs = null)
    {
        var pr = await Gh.GetPrAsync(config.Owner, config.Repo, prNumber);

        if (pr.TryGetProperty("merged", out var merged) && merged.GetBoolean())
            return new PrAnalysis
            {
                Number = prNumber, Title = pr.GetProperty("title").GetString()!,
                HeadSha = pr.GetProperty("head").GetProperty("sha").GetString()!,
                HeadRef = pr.GetProperty("head").GetProperty("ref").GetString()!,
                Status = PrStatus.Merged,
            };

        if (pr.GetProperty("state").GetString() == "closed")
            return new PrAnalysis
            {
                Number = prNumber, Title = pr.GetProperty("title").GetString()!,
                HeadSha = pr.GetProperty("head").GetProperty("sha").GetString()!,
                HeadRef = pr.GetProperty("head").GetProperty("ref").GetString()!,
                Status = PrStatus.Closed,
            };

        var analysis = new PrAnalysis
        {
            Number = prNumber,
            Title = pr.GetProperty("title").GetString()!,
            HeadSha = pr.GetProperty("head").GetProperty("sha").GetString()!,
            HeadRef = pr.GetProperty("head").GetProperty("ref").GetString()!,
            Mergeable = pr.TryGetProperty("mergeable_state", out var ms)
                        && ms.GetString() == "clean",
            HasMergeConflicts = pr.TryGetProperty("mergeable_state", out var ms2)
                                && ms2.GetString() == "dirty",
        };

        CheckSuperseded(config, analysis, pr, allOpenPrs);
        if (analysis.IsSuperseded)
        {
            analysis.Status = PrStatus.Superseded;
            return analysis;
        }

        await CheckCiAsync(config, analysis);
        await CheckReviewsAsync(config, analysis);

        // Detect merge conflicts as a fixable issue
        if (analysis.HasMergeConflicts)
            analysis.FixableIssues.Insert(0, new FixableIssue(
                "Merge conflicts with base branch (will resolve using newest versions)",
                "", "merge_conflicts"));

        // Fixable issues take priority -- the CI failure may be caused by them
        if (analysis.FixableIssues.Count > 0)
            analysis.Status = PrStatus.HasReviewIssues;
        else if (analysis.CiPending)
            analysis.Status = PrStatus.WaitingForCi;
        else if (!analysis.CiPassed)
            analysis.Status = PrStatus.CiFailing;
        else if (analysis.HasApprovals)
            analysis.Status = PrStatus.ReadyToMerge;
        else
            analysis.Status = PrStatus.ReadyToApprove;

        return analysis;
    }

    private static void CheckSuperseded(
        AgentConfig config, PrAnalysis analysis, JsonElement pr,
        List<JsonElement>? allOpenPrs)
    {
        if (allOpenPrs is null) return;

        var title = pr.GetProperty("title").GetString()!;
        foreach (var pattern in config.PrTitlePatterns)
        {
            if (!title.Contains(pattern)) continue;

            var newer = allOpenPrs
                .Where(p => p.GetProperty("title").GetString()!.Contains(pattern)
                            && p.GetProperty("number").GetInt32() > analysis.Number)
                .OrderByDescending(p => p.GetProperty("number").GetInt32())
                .FirstOrDefault();

            if (newer.ValueKind != JsonValueKind.Undefined)
            {
                analysis.IsSuperseded = true;
                analysis.NewerPrNumber = newer.GetProperty("number").GetInt32();
                Log.Info($"PR #{analysis.Number} is superseded by #{analysis.NewerPrNumber}");
            }
            break;
        }
    }

    private static async Task CheckCiAsync(AgentConfig config, PrAnalysis analysis)
    {
        var checks = await Gh.GetCheckRunsAsync(config.Owner, config.Repo, analysis.HeadSha);
        if (checks.Count == 0) { analysis.CiPending = true; return; }

        var mainCheck = checks.FirstOrDefault(c =>
            c.GetProperty("name").GetString() == config.RequiredCheckPattern);

        if (mainCheck.ValueKind != JsonValueKind.Undefined)
        {
            var status = mainCheck.GetProperty("status").GetString();
            var conclusion = mainCheck.TryGetProperty("conclusion", out var cc) ? cc.GetString() : null;

            if (status != "completed")
                analysis.CiPending = true;
            else if (conclusion == "success")
                analysis.CiPassed = true;
            else
            {
                analysis.CiPassed = false;
                analysis.CiFailedChecks.Add($"{mainCheck.GetProperty("name").GetString()}: {conclusion}");
            }
        }
        else
        {
            var completed = checks.Where(c => c.GetProperty("status").GetString() == "completed").ToList();
            var pending = checks.Where(c => c.GetProperty("status").GetString() != "completed").ToList();
            var failed = completed.Where(c =>
                c.TryGetProperty("conclusion", out var cc) && cc.GetString() != "success").ToList();

            if (pending.Count > 0) analysis.CiPending = true;
            else if (failed.Count > 0)
            {
                analysis.CiPassed = false;
                foreach (var c in failed)
                {
                    var name = c.GetProperty("name").GetString();
                    var conclusion = c.TryGetProperty("conclusion", out var cc2) && cc2.GetString() is string s ? s : "unknown";
                    analysis.CiFailedChecks.Add($"{name}: {conclusion}");
                }
            }
            else analysis.CiPassed = true;
        }
    }

    private static async Task CheckReviewsAsync(AgentConfig config, PrAnalysis analysis)
    {
        var reviews = await Gh.GetPrReviewsAsync(config.Owner, config.Repo, analysis.Number);
        var comments = await Gh.GetPrReviewCommentsAsync(config.Owner, config.Repo, analysis.Number);

        analysis.ApprovalCount = reviews.Count(r =>
            r.GetProperty("state").GetString() == "APPROVED");
        analysis.HasApprovals = analysis.ApprovalCount > 0;

        foreach (var comment in comments)
        {
            var body = comment.GetProperty("body").GetString() ?? "";
            var path = comment.TryGetProperty("path", out var p) ? p.GetString() ?? "" : "";
            ClassifyReviewComment(analysis, body, path);
        }
    }

    private static void ClassifyReviewComment(PrAnalysis analysis, string body, string path)
    {
        var lower = body.ToLowerInvariant();

        if (lower.Contains("package-lock") &&
            (lower.Contains("stale") || lower.Contains("still pins") ||
             lower.Contains("regenerate") || lower.Contains("update")))
        {
            analysis.FixableIssues.Add(new FixableIssue(
                "package-lock.json is out of sync with package.json",
                path, "lockfile_stale"));
            return;
        }

        if ((lower.Contains("bump") || lower.Contains("version")) &&
            (lower.Contains("mismatch") || lower.Contains("still") || lower.Contains("please")))
        {
            var match = VersionRegex().Match(body);
            analysis.FixableIssues.Add(new FixableIssue(
                $"Version needs updating in {path}", path, "version_mismatch",
                new() { ["suggested_version"] = match.Success ? match.Groups[1].Value : null }));
            return;
        }

        if (lower.Contains("devdependencies") || lower.Contains("dev dependencies"))
        {
            analysis.FixableIssues.Add(new FixableIssue(
                "devDependencies version needs bumping", path, "version_mismatch"));
            return;
        }

        if (body.Length > 50)
            analysis.UnfixableIssues.Add($"[{path}] {body[..Math.Min(200, body.Length)]}");
    }

    [GeneratedRegex(@"(\d+\.\d+\.\d+(?:-[a-zA-Z0-9.]+)?)")]
    private static partial Regex VersionRegex();
}
